src/main/java/org/kuali/kra/negotiations/lookup/NegotiationDaoOjb.java

```diff
@@ -28,394 +30,247 @@
+        LOG.debug("getNegotiationResults fieldValues={}",fieldValues);
+        CollectionIncomplete<Negotiation> searchResults = new CollectionIncomplete<>(new ArrayList(0), 0L);
+
+        //negotiation Age filtering done as a special case
+        negotiationAge = fieldValues.get(NEGOTIATION_AGE);
+        fieldValues.remove(NEGOTIATION_AGE);
+
+        Map<String, String> associationDetails = getAssociationFilters( fieldValues );
+
+
+        //add additional criteria related to associated objects - if any
         if (!associationDetails.isEmpty()) {
-            addListToList(result, getNegotiationsLinkedToAward(fieldValues, associationDetails));
-            addListToList(result, getNegotiationsLinkedToProposal(fieldValues, associationDetails));
-            addListToList(result, getNegotiationsLinkedToProposalLog(fieldValues, associationDetails));
-            addListToList(result, getNegotiationsUnassociated(fieldValues, associationDetails));
-            addListToList(result, getNegotiationsLinkedToSubAward(fieldValues, associationDetails));
+            addListToList(searchResults, getNegotiationsLinkedToAward(fieldValues, associationDetails));
+            addListToList(searchResults, getNegotiationsLinkedToProposal(fieldValues, associationDetails, searchResults.size()>=getMaxSearchResultLimit()));
+            addListToList(searchResults, getNegotiationsUnassociated(fieldValues, associationDetails, searchResults.size()>=getMaxSearchResultLimit()));
+            addListToList(searchResults, getNegotiationsLinkedToSubAward(fieldValues, associationDetails, searchResults.size()>=getMaxSearchResultLimit()));
+
         } else {
-            result = findCollectionBySearchHelper(Negotiation.class, fieldValues, false, false);
+            //generate negotiation query criteria from user filters
+            Criteria negotiationSearchCriteria = getCollectionCriteriaFromMap(new Negotiation(), fieldValues);
+            LOG.debug("Negotiation search criteria: {}", negotiationSearchCriteria);
+            searchResults = executeSearch(negotiationSearchCriteria, false);
         }
-        if (result != null && !result.isEmpty() && StringUtils.isNotBlank(fieldValues.get("negotiationAge"))) {
-            try {
-                result = filterByNegotiationAge(fieldValues.get("negotiationAge"), result);
-            } catch (NumberFormatException e) {
-                GlobalVariables.getMessageMap().putError(KRADConstants.DOCUMENT_ERRORS, RiceKeyConstants.ERROR_CUSTOM, new String[] { "Invalid Numeric Input: " + fieldValues.get("negotiationAge")});
-                result = new ArrayList<Negotiation>();
-            }
-        } return result;
-    }
-
-    /*
-     * used to filter the Negotitations returned from getNegotiationResults
-     */
-    private Collection<Negotiation> filterNegotitations(Collection<Negotiation> negotiations,String title,String unitNumber,String unitName,String requisitionerId)
```

Comment:
Wow, this was ugly!

Reply:
Yes, muy feo! :+1: This is the second attempt at refactoring... first one failed miserably